### PR TITLE
Don't require TLS for in-process connection

### PR DESCRIPTION
**Pull request by neilalexander**
This should fix a bug where in-process connections expect TLS over the `net.Pipe` if TLS is configured.

Signed-off-by: Neil Twigg <neil@nats.io>

### CODE REVIEW

**Comment by neilalexander** [comment]
@caleblloyd Please test against this commit, should fix your issue.

File: server/server.go [hunk 1]
```diff
@@ -2571,7 +2571,7 @@ func (c *tlsMixConn) Read(b []byte) (int, error) {
 	return c.Conn.Read(b)
 }
 
-func (s *Server) createClient(conn net.Conn) *client {
+func (s *Server) createClient(conn net.Conn, inProcess bool) *client {
```
Comment (derekcollison) [inline server/server.go:2574]: Prefer a new func called createInProcessClient(), keep the create client the way it is and add in another that takes the bool flag..

Comment (derekcollison) [reply server/server.go:2574]: We will be formalizing alot on clients, so might be worth it imo.

Comment (neilalexander) [reply server/server.go]: Sure, have done.

File: server/server.go [hunk 2]
```diff
@@ -2572,6 +2572,14 @@ func (c *tlsMixConn) Read(b []byte) (int, error) {
...
 
 func (s *Server) createClient(conn net.Conn) *client {
+	return s._createClient(conn, false)
+}
+
+func (s *Server) createClientInProcess(conn net.Conn) *client {
+	return s._createClient(conn, true)
+}
+
+func (s *Server) _createClient(conn net.Conn, inProcess bool) *client {
```
[the remark this answers is not in the document]
Comment (neilalexander) [reply server/server.go]: Fixed

**Review by derekcollison — APPROVED** [review]
LGTM

**Comment by caleblloyd** [comment]
@sethjback can you test the latest commit to make sure TLS is not required now?

**Comment by caleblloyd** [comment]
Getting a new error.  Before this change, was seeing:

```
Error: tls: failed to verify certificate: x509: certificate signed by unknown authority
```

After this change, am seeing:

```
Error: tls: first record does not look like a TLS handshake
```

**Comment by derekcollison** [comment]
I will wait to merge until I hear from @neilalexander on the issue above.

**Comment by neilalexander** [comment]
Have pushed another commit that should resolve this without having to set `AllowNonTLS` in the config file, which would apply to all connections. Please give it a try.

**Comment by caleblloyd** [comment]
`tls_required: true` is still being sent in Server Info to the client, so the Client is attempting to upgrade to TLS...  Think we need to detect/change that for the in-process client to send to `tls_required: false, tls_available: true`

**Comment by caleblloyd** [comment]
I think https://github.com/nats-io/nats-server/commit/4b064ff8eaf90346372c137927247eeb889a40c0 works, feel free to cherry-pick that as it is based on this branch already

**Comment by neilalexander** [comment]
OK, have added a unit test this time.

**Review by caleblloyd — APPROVED** [review]
Looks good!

File: test/tls_test.go [hunk 3]
```diff
@@ -72,6 +72,26 @@ func TestTLSConnection(t *testing.T) {
 	}
 }
 
+// TestTLSInProcessConnection checks that even if TLS is enabled on the server,
+// that an in-process connection that does *not* use TLS still connects successfully.
+func TestTLSInProcessConnection(t *testing.T) {
+	srv, opts := RunServerWithConfig("./configs/tls.conf")
+	defer srv.Shutdown()
+
+	nc, err := nats.Connect("", nats.InProcessServer(srv), nats.UserInfo(opts.Username, opts.Password))
```
Comment (derekcollison) [inline test/tls_test.go:81]: Small nit, _EMPTY_ vs ""

**Review by derekcollison — APPROVED** [review]
LGTM

File: test/tls_test.go (hunk 3 above)
Comment (derekcollison) [reply test/tls_test.go:81]: Let me know if you fix or not and will merge after.

Comment (neilalexander) [reply test/tls_test.go:81]: `_EMPTY_` isn't defined in the `test` package, only in `server`.

**Pull request merged**